src/util/beacon.ts

```diff
@@ -1,17 +1,43 @@
-import { hasProperty } from "./object";
+import { SendBeaconResult } from "../@types"
 
-const hasBeaconSupport = hasProperty(navigator, "sendBeacon");
+/**
+ * A module-level variable caching the result of `sendBeacon` feature
+ * detection.
+ */
+const hasBeaconSupport = "sendBeacon" in navigator
 
-function beacon(url: string, data: string): void {
-  if (hasBeaconSupport) {
-    navigator.sendBeacon(url, data);
-  } else {
-    fetch(url, {
-      method: "POST",
-      body: data,
-      keepalive: true
-    });
-  }
+/**
+ * Sends beacon data.
+ *
+ * @remarks
+ *
+ * Uses the Beacon API if available. Otherwise, sends a POST message using the
+ * Fetch API.
+ *
+ * ### References
+ *
+ * On MDN:
+ * * [Beacon API](https://developer.mozilla.org/en-US/docs/Web/API/Beacon_API)
+ * * [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API)
+ *
+ * @param url URL to send beacon data to
+ * @param data Beacon data to send
+ */
+export default function beacon(
+    url: string,
+    data: string,
+): Promise<SendBeaconResult> {
+    if (hasBeaconSupport) {
+        if (navigator.sendBeacon(url, data)) {
+            return Promise.resolve()
+        }
+        // TODO: should we attempt to use fetch in this case, or does that risk
+        // duplicate reporting?
```

Comment:
Great question. Looking at the [spec processing model](https://www.w3.org/TR/beacon/#h-sec-processing-model) it seems pretty safe for us to fall through this block to the fetch below. And something i've not considered before 👍